### DISCUSSION TITLE
Missing BLE config params

### DISCUSSION THREAD

**Discussion by Wojciech-Bie** [micropython/micropython#15792] [RP2040 / Pico]
Is there a plan to implement BLE config params like 'bond', 'le_secure', 'mitm', 'io'? I don't see them in [v1.23.0](https://micropython.org/resources/firmware/RPI_PICO-20240602-v1.23.0.uf2)

**Comment by andrewleech** [comment]
This is a much bigger topic than just the config params - Pico uses btstack to provide BLE support, but pairing and bonding integration into micropython hasn't been finished yet for btstack. Other ports like stm32 / esp32 with pairing are based on the nimble stack instead. 

@brianreinhold has been putting a huge amount of work into this space and making a lot of progress, but it's a surprisingly hard thing to get working.

**Comment by brianreinhold** [comment]
Yes, we have done a lot of work to add pairing support for the PICO-W with the btstack with aioble and we have a working implementation, at least for the client side. We had to do this given our customer requirements. It required significant modifications to `modbluetooth_btstack.c` and aioble. Aioble also does not support persisting service discovery which is problematic for bonded reconnects. Working on that now and have a solution (needs further testing and refinement). Aioble also has issues when the server unexpectedly disconnects - which they do unfortunately. Handling this issue required a direct communication between the application and the IRQ disconnect and connect events in aioble in order to prevent the application from doing something (like service discovery or reading the DIS) when the device has disconnected. By the time aioble signals the application of the disconnect, the application would otherwise try to invoke some action it should not which causes aioble to trigger one of many possible exceptions, often a value error.

However, it's will take someone who is much more knowledgeable about Micropython and its build environment and aioble than I am to generalize the work and make it useful for all. Again, we have only implemented what was necessary to support a PICO client though the server side should probably be simpler.

**Comment by Wojciech-Bie** [comment]
Thanks for clearing things out. @brianreinhold is your implementation available somewhere? If I understand correctly, you don't feel a need to integrate your work to Micropython, is that right?

**Comment by brianreinhold** [comment]
@Wojciech-Bie Its not that I don't feel the need but that I lack the knowledge and expertise to do so. I have limited experience with Python in any form and limited understanding of the Micropython build package. I have done a lot of work with Bluetooth Low Energy on Android (client) and the Nordic semiconductor platforms (peripheral) and have worked on B-SIG standards development and know what a Bluetooth library needs to do. I am sure I have done many things that would be unacceptable if I were to make a pull request, especially with respect to the aioble library. And as I said, I have only done it for the client side.
